// react-light-notifications/lib/components/NotificationManager.js
'use strict';

exports.__esModule = true;

var _events = require('events');

var _v = require('uuid/v4');

var _v2 = _interopRequireDefault(_v);

var _types = require('../constants/types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NotificationManager = function (_EventEmitter) {
  _inherits(NotificationManager, _EventEmitter);

  function NotificationManager() {
    _classCallCheck(this, NotificationManager);

    var _this = _possibleConstructorReturn(this, _EventEmitter.call(this));

    _this.queue = [];
    return _this;
  }

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {function} options.onClick
   * @param {string} type
   */


  NotificationManager.prototype.create = function create(options, type) {
    var defaultOptions = {
      id: (0, _v2.default)(),
      type: type || _types.TYPES.INFO,
      title: null,
      message: null,
      timeOut: 5000
    };

    if (options && options.priority) {
      this.queue.unshift(Object.assign(defaultOptions, options));
    } else {
      this.queue.push(Object.assign(defaultOptions, options));
    }

    this.emitChange();
  };

  NotificationManager.prototype.checkOptions = function checkOptions(input) {
    if (typeof input === 'string') return { message: input };else return input;
  };

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {string} options.position
   * @param {function} options.onClick
   */


  NotificationManager.prototype.info = function info(options) {
    this.create(this.checkOptions(options), _types.TYPES.INFO);
  };

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {string} options.position
   * @param {function} options.onClick
   */


  NotificationManager.prototype.success = function success(options) {
    this.create(this.checkOptions(options), _types.TYPES.SUCCESS);
  };

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {string} options.position
   * @param {function} options.onClick
   */


  NotificationManager.prototype.warning = function warning(options) {
    this.create(this.checkOptions(options), _types.TYPES.WARNING);
  };

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {string} options.position
   * @param {function} options.onClick
   */


  NotificationManager.prototype.error = function error(options) {
    this.create(this.checkOptions(options), _types.TYPES.ERROR);
  };

  /**
   *
   * @param {object} options
   * @param {string} options.title
   * @param {string} options.message
   * @param {bool} options.showCloseButton
   * @param {bool} options.timeOut
   * @param {bool} options.priority
   * @param {function} options.onClick
   * @param {function} options.bgColor
   * @param {function} options.iconClassName
   */


  NotificationManager.prototype.custom = function custom(options) {
    this.create(this.checkOptions(options), _types.TYPES.CUSTOM);
  };

  NotificationManager.prototype.remove = function remove(notification) {
    this.queue = this.queue.filter(function (n) {
      return notification.id !== n.id;
    });
    this.emitChange();
  };

  NotificationManager.prototype.emitChange = function emitChange() {
    this.emit(_types.TYPES.CHANGE, this.queue);
  };

  NotificationManager.prototype.addChangeListener = function addChangeListener(callback) {
    this.addListener(_types.TYPES.CHANGE, callback);
  };

  NotificationManager.prototype.removeChangeListener = function removeChangeListener(callback) {
    this.removeListener(_types.TYPES.CHANGE, callback);
  };

  return NotificationManager;
}(_events.EventEmitter);

exports.default = new NotificationManager();
module.exports = exports['default'];
